### ISSUE TITLE
Add support for gRPC

### ISSUE THREAD

**Issue by jwulf** [denoland/deno#3326]
Is it possible to write a gRPC client application in Deno?

**Comment by benjamingr** [comment]
Sure, what do you mean?

**Comment by ry** [comment]
There is no gRPC module yet.

**Comment by tuddman** [comment]
would this be the way to go? https://github.com/stepancheg/grpc-rust

**Comment by sleipnir** [comment]
Hello @tuddman , I understand that Deno is written over Rust but I think the question is about how I use gRPC in Deno, how do I write a gRPC client or server code in Deno? It's possible? Is there a library ready to use? If not, how to use the link you sent to provide a Javascript / Typescript api at the user level?

**Comment by jadhavajay** [comment]
Looking forward to the possibility of using gRPC with Deno.

**Comment by sleipnir** [comment]
I believe that due to the popularity of gRPC it would be great to see support in the Deno standard library, like any other form of IO

**Comment by theoutlander** [comment]
I would like to help with this integration, but I'm wondering if the best option is integration with grpc-rust or writing a typescript library? What about porting something like https://github.com/grpc/grpc-web/tree/master/packages/grpc-web as a typescript lib?

**Comment by kitsonk** [comment]
IMO, this is something that should start external to Deno/`std` and move into `std`.  It should either be a native module or Web Assembly.

**Comment by badsyntax** [comment]
It should be noted the current [advised approach](https://grpc.io/blog/grpc-js-1.0/) of using gRPC in Node.js is to use the [pure JavaScript module](https://www.npmjs.com/package/@grpc/grpc-js) which uses Node.js's built-in modules (http2) instead of a native module built in c++. The module is [built in TypeScript](https://github.com/grpc/grpc-node/tree/master/packages/grpc-js) and perhaps can be used for inspiration for something in Deno, assuming [http2 support](https://github.com/denoland/deno/issues/3995) is provided by Deno at some point. cc @murgatroid99

**Comment by murgatroid99** [comment]
Hi. I'm the author of the existing gRPC libraries for Node.js. I think that porting `@grpc/grpc-js` to Deno might be the way to go, but that would depend on how close Deno's eventual HTTP/2 implementation is to Node's `http2` module, semantically. Other built in modules may also be a concern, particularly `tls`.

I would not suggest porting grpc-web. That library implements a different protocol to account for the restrictions of browser APIs. It needs a proxy to talk to regular gRPC servers.

**Comment by sleipnir** [comment]
I aggre about grpc-web maybe this should be another Deno module

**Comment by yurianxdev** [comment]
@murgatroid99 I have a doubt, how is gRPC-node generating code from Proto right now?

**Comment by murgatroid99** [comment]
There are two major methods: one is to use the `@grpc/proto-loader` package to load `.proto` files and generate classes and object at runtime. That uses Protobuf.js internally.

The alternative is to use the `protoc` tool distributed with a plugin in the `grpc-tools` package to generate JavaScript files that depend on the `google-protobuf` package.

I want to note that neither of these is strictly necessary in general to use `grpc`. The gRPC protocol (and the library) can handle any kinds of serializable messages.

**Comment by yurianxdev** [comment]
Ok, let me understand a little bit. So first we need to implement something like grpc-deno, which should have support for any kind of serializable message. Then adapting (or build a new one) the tooling for generating grpc code that uses grpc-deno as a reference, from proto (or any other) to TypeScript/Javascript.

**Comment by murgatroid99** [comment]
That looks about right.

One thing I want to note is that there is a kind of generic interchange format for defining a collection of services that `grpc.loadPackageDefinition` accepts and `@grpc/proto-loader` generates, and there is also an option to have `grpc-tools` generate it. That interchange format is defined as `PackageDefinition` in [this document](https://github.com/grpc/proposal/blob/master/L23-node-protobufjs-library.md). I think it would be good if any implementation for Deno can load the same kind of objects. 

Ideally I think gRPC for Deno should have the same API and functionality as the existing `@grpc/grpc-js`, but I recognize that that may not be feasible.

**Comment by yurianxdev** [comment]
@murgatroid99 there is an article or a document explaining how grpc-node (or other C-grpc-based libraries) uses the core-grpc implementation?

**Comment by yurianxdev** [comment]
Another thing @murgatroid99, What do you think about building a grpc-deno on top of [tonic](https://github.com/hyperium/tonic) (pure Rust implementation of grpc).

**Comment by murgatroid99** [comment]
I'm not aware of such a public document about using the gRPC core library. We did most of that work before we started doing our design work publicly. The simple answer, though, is that we defined a public API for JavaScript users (in this case), and then an intermediate API, that the implementation of the public API could consume and that we could reasonably implement in C++ using the core API. That intermediate API is considered non-public and not guaranteed to be stable. If you look at the `grpc` package directories, the `ext/` directory contains the implementation of that intermediate API and the `src/` directory contains the implementation of the public API.

The Node library in particular is unusual in that the gRPC core library has a libuv implementation of the low-level network interaction that was created specifically for the Node library, which can be enabled using a compiler flag. So the Node gRPC library enables that to simplify its interaction with the event loop.

I think it's important to mention that the native addon was a maintainability nightmare, to the point that we decided to reimplement the library purely in TypeScript as `@grpc/grpc-js` and we are phasing out the use of the native addon-based library in favor of the pure TypeScript implementation.

Regarding implementing this on top of tonic, I think the important question is how close the desired JavaScript API is to the tonic API, semantically. In addition, how much extra work will it be to maintain the Deno equivalent of a native addon using a Rust library? And how does the total implementation work there compare with modifying `@grpc/grpc-js` for use with Deno, once the necessary standard libraries are available.

**Comment by yurianxdev** [comment]
I have taken a look at your pure TS implementation of grpc, and you are using only a single node dependency, so a deno library couldn't be hard based on that implementation? Based on what you said about the core library could be a nightmare in the long term, even with a Rust implementation.

**Comment by theoutlander** [comment]
I like the typescript based approach. Maybe we need to implement http/2
first, then it is mostly straightforward.

**Comment by yurianxdev** [comment]
There is no progress on HTTP2 (#3995) so if we want to follow the pure TS approach we need to implement it. That will kill two birds with one shot, I like it.

**Comment by kitsonk** [comment]
HTTP2 will eventually land.

Again, this needs to start as something seperate from `std` or the Deno cli.

**Comment by jd1378** [comment]
any news on this ?

**Comment by anonrig** [comment]
Hi! I'm a maintainer from mali (GRPC microservice framework) and wanted to stay in the loop. We're currently investigating supporting Deno in Mali. If you're interested in this support, please send us a message from https://github.com/malijs/mali/issues/241

**Comment by brunoluiz** [comment]
Seems Deno 1.9, released today, implements HTTP 2.0 🎉 

https://deno.com/blog/v1.9#native-http%2F2-web-server

**Comment by murgatroid99** [comment]
The post says that it supports full duplex communication. That's good. Does it also have full support for trailers, on the client side and the server side? That is important for a gRPC implementation.

**Comment by prohazko2** [comment]
Was playing around with gRPC implementation and almost succeed with basic unary calls serving

1. http2 with `Deno.serveHttp` was ok
2. protobuf msgs decoding/encoding with `protobuf.js` was ok
3. thing that I didn't figure out - how to send `OK` status back to client

Similar node's `@grpc/grpc-js` calls has two server trailers: 
```
[0] Received server trailers:
		grpc-status: 0
		grpc-message: OK

[0] received status code 0 from server
[0] received status details string "OK" from server
```
Seems like they are sent with this line of code from
[@grpc/grpc-js@1.3.x ... server-call.ts](https://github.com/grpc/grpc-node/blob/%40grpc/grpc-js%401.3.x/packages/grpc-js/src/server-call.ts#L602)

But I didn't have any luck with sending them with Deno's `respondWith` / `Response` API's

https://github.com/prohazko2/deno-grpc/blob/0.1.3/serve.ts

**Comment by bartlomieju** [comment]
@prohazko2 trailing headers are not currently supported. Could you open a new issue with your use case? We'll try to look into it soon.

**Comment by prohazko2** [comment]
@bartlomieju sure, here it goes https://github.com/denoland/deno/issues/10214

**Comment by prohazko2** [comment]
Hi,
If anyone still interested - i made some progress on my lib https://github.com/prohazko2/deno-grpc
It now can make and serve unary calls.
Main bummer is - it doesn't use Deno's HTTP/2 bindings but uses JS impl from  now deprecated https://github.com/molnarg/node-http2 repo.
Also not expect much from it - I do it only for fun and have no intentions on implementing full gRPC spec

**Comment by badsyntax** [comment]
As per https://github.com/denoland/deno/issues/10214#issuecomment-821723392

> I don't think we will be supporting them either if the web does not.

Does this effectively mean we cannot port [grpc-js](https://github.com/grpc/grpc-node/tree/master/packages/grpc-js) to deno?

**Comment by benatkin** [comment]
@badsyntax no, it means that it isn't going to be in core.

I think it might be implemented in an ideal fashion as an ffi feature, and that we might need an `--allow-ffi` allow list or something.

**Comment by stabai** [comment]
At the risk of beating a dead horse, there is an unobvious use cases that Deno would be awesome for, but also depend on gRPC support. Plugins for Terraform, Aspect, and other tools that use this pattern can be [written in any framework that supports gRPC](https://github.com/hashicorp/go-plugin/blob/master/README.md#go-plugin-system-over-rpc). For developers who work with gRPC, this also means interoperability with all of their own tools/services in their development platform. The current necessity to do these things in Node or some other language is far from ideal, especially for those who are trying to adopt Deno.

**Comment by stabai** [comment]
@benatkin 
> I think it might be implemented in an ideal fashion as an ffi feature, and that we might need an `--allow-ffi` allow list or something.

Is there another issue tracking that for us to follow?

**Comment by kitsonk** [comment]
To give context, we believe that integrated gRPC to the Deno CLI is an important feature and are committing core team resources to delivering it.

**Comment by johynpapin** [comment]
Support for gRPC in the Deno CLI will be an extremely nice feature, thank you for choosing to do so!
I think that supporting the [Connect protocol](https://connect.build/docs/protocol) (with gRPC and gRPC-Web) would be interesting.
Is it possible to participate in the development of gRPC for Deno?

**Comment by disjukr** [comment]
FYI, I posted a runtime API proposal for gRPC support in the discussions. https://github.com/denoland/deno/discussions/17308

**Comment by donoso-eth** [comment]
Hey, what is the status here? Can we expect support for gRPC ?

**Comment by satyarohith** [comment]
Hi @donoso-eth. We are actively working to support https://www.npmjs.com/package/@grpc/grpc-js in Deno. Right now, the following official examples work if you use es modules ([here is a branch](https://github.com/satyarohith/grpc-node/tree/es_examples) with updated examples)

- helloworld
- cancellation
- interceptors        
- deadline        
- metadata
- keepalive

The next steps are to get the error_handling example working and popular packages that rely on gRPC.

**Comment by hronro** [comment]
It's very cool to know grpc-js will be available on Deno soon, I would also like to know what the performance it would be.

Since Deno already has a very performant built-in HTTP server, I would expect Deno will eventually have a gRPC implementation at the same performance level. I also wonder if it's possible to achieve that performance via pure JavaScript, or if some parts must be written in Rust.

**Comment by birkskyum** [comment]
So the blog post make it look like everything is working. Is that the case?

https://deno.com/blog/v1.44#grpc-connections-now-supported

<img width="764" alt="Screenshot 2024-05-30 at 17 12 24" src="https://github.com/denoland/deno/assets/74932975/cc6b7ef6-bd73-43fc-ae48-d97397fd75e5">


Related
- #23714 
- #23246

**Comment by satyarohith** [comment]
@birkskyum we have support for all of the official gRPC client side examples and we also tested google packages like vision that rely on gRPC client APIs and they are working. We don't support the server side yet.

**Comment by xiaojiudev** [comment]
@rnbguy @satyarohith  I am using grpc_basic, but it's getting an error, and the author of that library is no longer providing support. For the grpc_basic example in the server, it needs Deno to listen on a port, and the grpc_server will handle that port (Deno.TCPConn), as shown in the image below:
![image](https://github.com/denoland/deno/assets/89904184/3c89f2fc-fe48-4d88-8f5e-efcd38b5dd9a)

Now, as you say, Deno v1.44 now support grpc/grpcjs. so I want to implement grpc/grpc-js, but I have a small issue. The library needs bindAsync directly on the port, and there's no longer a need for using Deno.listen. However, **I want to know if it's possible to use Deno.listen instead?**
![image](https://github.com/denoland/deno/assets/89904184/ccd9da06-da67-40a7-9215-17e0a1e9b8a9)

**Comment by rnbguy** [comment]
hey @xiaojiudev, I am not really part of the Deno team. But AFAIK, the current Deno `@grpc/grpc-js` support is only for gRPC clients.

**Comment by niteshbalusu11** [comment]
I am trying to get a grpc client working in deno. I used protobufjs-cli to generate a JS and d.ts file.
When importing the file and running in deno I am getting the following error:

```
error: The module's source code could not be parsed: 'import', and 'export' cannot be used outside of module code at file:///Users/nitesh/Documents/dunder-lsp-lsps/src/proto.mjs:19:5

      export const lnrpc = $root.lnrpc = (() => {
```      

Also inside the generated JS file, getting this where it says `export const lnrpc`

```
Modifiers cannot appear here.
```
Generated JS file looks something like this:

```js
/*eslint-disable block-scoped-var, id-length, no-control-regex, no-magic-numbers, no-prototype-builtins, no-redeclare, no-shadow, no-var, sort-vars*/
(function(global, factory) { /* global define, require, module */

    /* AMD */ if (typeof define === 'function' && define.amd)
        define(["protobufjs/minimal"], factory);

    /* CommonJS */ else if (typeof require === 'function' && typeof module === 'object' && module && module.exports)
        module.exports = factory(require("protobufjs/minimal"));

})(this, function($protobuf) {
    "use strict";

    // Common aliases
    const $Reader = $protobuf.Reader, $Writer = $protobuf.Writer, $util = $protobuf.util;
    
    // Exported root namespace
    const $root = $protobuf.roots["default"] || ($protobuf.roots["default"] = {});
    
    export const lnrpc = $root.lnrpc = (() => {
    
        /**
         * Namespace lnrpc.
         * @exports lnrpc
         * @namespace
         */
        const lnrpc = {};
    
        lnrpc.Lightning = (function() {
    
            /**
             * Constructs a new Lightning service.
             * @memberof lnrpc
             * @classdesc Represents a Lightning
             * @extends $protobuf.rpc.Service
             * @constructor
             * @param {$protobuf.RPCImpl} rpcImpl RPC implementation
             * @param {boolean} [requestDelimited=false] Whether requests are length-delimited
             * @param {boolean} [responseDelimited=false] Whether responses are length-delimited
             */
            function Lightning(rpcImpl, requestDelimited, responseDelimited) {
                $protobuf.rpc.Service.call(this, rpcImpl, requestDelimited, responseDelimited);
            }
    
    ```

**Comment by garethj2** [comment]
I'm running Deno 1.44.4 and have a client working. Unfortunately everything seems to stop working after a minute. I've run exactly the same code using NodeJS and there is no problem.

My code simply opens a stream and listens for events.

Unfortunately I don't have any errors I can report from the output.

**Comment by satyarohith** [comment]
@garethj2 the issue you [reported](https://github.com/denoland/deno/issues/24305) is fixed in https://github.com/denoland/deno/pull/24576

**Comment by butadpj** [comment]
I just submitted a new issue - #26436 that seems related to Deno's lack of support for server-side gRPC.

**Comment by satyarohith** [comment]
Closing this thread as the original issue is about [gRPC client applications](https://github.com/denoland/deno/issues/3326#issue-521966885) which are now supported. Please open new issues for bugs related to clients. And support for server side is being tracked at #23714
